### ISSUE TITLE
Preemptive proxy authentication

### ISSUE THREAD

**Issue by smatei** [square/okhttp#2435]
Hi,

Is there a way to make a basic proxy authentication preemptive? By default, the request is done in RealCall.getResponse - engine.sendRequest(), then HttpEngine.followUpRequest is called that detects responseCode HTTP_PROXY_AUTH and then the authentication is done

return client.proxyAuthenticator().authenticate(route, userResponse);

(https://github.com/square/okhttp/pull/2417, https://github.com/square/okhttp/pull/2414)

Then another sendRequest is called that is successful.

Is there a way to force authenticate from the first sendRequest? Because I know my proxy needs authentication, I don't need the server to tell me this.

In apache http client I can set authentication preemptive. 

aClient.getParams().setAuthenticationPreemptive(true);

Do  you have such a thing?

Stefan Matei

**Comment by swankjesse** [comment]
Nothing at the moment. We should offer something here.

**Comment by smatei** [comment]
Please add it as a feature request.

Stefan Matei

**Comment by smatei** [comment]
Hi, 

Figured out how to authenticate directly. Added the "Proxy-Authorization" header to my original request and everything is working perfect.

`String credential = Credentials.basic(user, password);
    
    Request request = new Request.Builder()
        .url(url)
        .header("User-Agent", "Mozilla/5.0 (Windows NT 6.2; WOW64; rv:43.0) Gecko/20100101 Firefox/43.0")
        .header("Proxy-Authorization", credential)
        .build();`

**Comment by swankjesse** [comment]
Considering the APIs for this. We already have this:
```
interface Authenticator {
  Request authenticate(Route route, Response response);
```
I’m tempted by this:
```
class AbstractAuthenticator implements Authenticator {
  Request authenticate(Request request);
  Request authenticate(Route route, Response response);
```
Certainly the name is terrible. But I like having both proactive and reactive authentication in one interface. And we can use it for both proxies and direct requests.

**Comment by jsoref** [comment]
@swankjesse: Did you mean `interface` for your `AbstractAuthenticator` as opposed to `class`?
ala:
`public interface AbstractAuthenticator extends Authenticator {`

**Comment by bsia** [comment]
Is it possible to do pre-emptive proxy authentication via interceptors currently?

**Comment by gupta11** [comment]
So using OkHttp how do we connect through a proxy on the wifi network? All suggestions I have tried so far have failed. I am using retrofit and okhttp3.

**Comment by nicktrav** [comment]
Bumping this.

We're running up against the same issue where we have a proxy that _always_ requires authentication. The first, unauthenticated request is therefore wasted, as it will always fail. We monitor failed auth attempts, so this is creating a lot of noise.

If I'm reading some of the previous comments correctly, it sounds like there is a workaround here in that an Interceptor can be used to set the header on the initial CONNECT request. We can probably use that for now, but having first-class support for "pre-emptive authentication" would be cool.

cc: @stfinney

**Comment by smatei** [comment]
@nicktrav 
I have implemented a PreemptiveAuthenticator on my fork - https://github.com/smatei/okhttp 

The solution was not accepted on the master branch, but it works for me in production perfectly.

**Comment by nicktrav** [comment]
@smatei - great! Any specific reason it wasn't accepted? Were there additional things that needed to be addressed before that could be merged in, perhaps?

**Comment by swankjesse** [comment]
Fix will be released in OkHttp 3.12.
